File: DSA/VaniAgrawal_590015846/Day_85_Question_3.py
# Your task: Implement component counting using traversal-based or union-find approach to identify and count all disconnected regions in undirected graph.
from collections import deque

def count_connected_components():
    vertices_str = input("Enter the number of vertices: ")
    num_vertices = int(vertices_str)
    edges_str = input("Enter the edges as a list of pairs (e.g., [[0,1],[1,2],[3,4]]): ")
    edges = eval(edges_str)

    adj = [[] for _ in range(num_vertices)]
    for u, v in edges:
        adj[u].append(v)
        adj[v].append(u)
    
    visited = [False] * num_vertices
    count = 0
    
    for i in range(num_vertices):
        if not visited[i]:
            count += 1
            queue = deque([i])
            visited[i] = True
            while queue:
                node = queue.popleft()
                for neighbor in adj[node]:
                    if not visited[neighbor]:
                        visited[neighbor] = True
                        queue.append(neighbor)
    
    print("Output:", count)

count_connected_components()

File: DSA/RATNA_KHANNA_590013141/DAY_85_QUESTION_3.c
#include <stdio.h>
#include <stdlib.h>

void dfs(int node, int v, int adj[v][v], int visited[v]) {
    visited[node] = 1;
    for (int i = 0; i < v; i++) {
        if (adj[node][i] && !visited[i]) {
            dfs(i, v, adj, visited);
        }
    }
}

int main() {
    int v, e;
    printf("Enter number of vertices: ");
    scanf("%d", &v);
    printf("Enter number of edges: ");
    scanf("%d", &e);

    int adj[v][v];
    for (int i = 0; i < v; i++) {
        for (int j = 0; j < v; j++) {
            adj[i][j] = 0;
        }
    }

    printf("Enter edges (u v):\n");
    for (int i = 0; i < e; i++) {
        int u, w;
        scanf("%d %d", &u, &w);
        adj[u][w] = 1;
        adj[w][u] = 1;
    }

    int visited[v];
    for (int i = 0; i < v; i++) visited[i] = 0;

    int count = 0;
    for (int i = 0; i < v; i++) {
        if (!visited[i]) {
            dfs(i, v, adj, visited);
            count++;
        }
    }

    printf("Number of Connected Components: %d\n", count);
    return 0;
}
